interactive_engine/executor/ir/runtime/src/router.rs

```diff
@@ -0,0 +1,59 @@
+//
+//! Copyright 2023 Alibaba Group Holding Limited.
+//!
+//! Licensed under the Apache License, Version 2.0 (the "License");
+//! you may not use this file except in compliance with the License.
+//! You may obtain a copy of the License at
+//!
+//! http://www.apache.org/licenses/LICENSE-2.0
+//!
+//! Unless required by applicable law or agreed to in writing, software
+//! distributed under the License is distributed on an "AS IS" BASIS,
+//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//! See the License for the specific language governing permissions and
+//! limitations under the License.
+
+use std::sync::Arc;
+
+use graph_proxy::apis::partitioner::{ClusterInfo, PartitionInfo};
+use graph_proxy::apis::ID;
+use graph_proxy::GraphProxyResult;
+
+pub type WorkerId = u64;
+
+/// A `Router` is used to route the data to the destination worker so that it can be properly processed,
+/// especially when the underlying data has been partitioned across the cluster.
+/// Given the partition information (by `PartitionInfo`) as well as how our cluster is managed (by `ClusterInfo`) and co-located with the graph data,
+/// we can implement the corresponding `route` function to guide the system to transfer the data to a proper destination worker.
+///
+/// For example, suppose our computer server contains 10 servers, each further forking 10 workers for processing queries.
+/// In addition, the graph is partitioned into these 10 servers by the following strategy:
+/// vertex of give ID is placed in the server with id i (0 to 9) i given ID % 10 == i, the vertex's adjacent edges are also placed with the vertex.
+/// Then the router can decide which worker should process the vertex of ID 25534 as follows:
+/// - It first do `25534 % 10 == 4`, which means it must be routed to the 4-th server.
+/// - Any worker in the 4-th server can process the vertex. Thus it randomly picks a worker, saying 5-th worker, which has ID 4 * 10 + 5 = 45.
+/// - Then 45-th worker will be returned for routing this vertex.
+pub trait Router: Send + Sync + 'static {
+    /// a route function that given the data, return the worker id that is going to do the query.
+    fn route(&self, data: &ID, job_workers: usize) -> GraphProxyResult<WorkerId>;
```

Comment:
Similarly, `ID` for data, not very well-designed. Consider:
```
pub trait RouteData {
    /// Comment this
    fn get_route_key(&self) -> ID; 
}

impl RouteData for Vertex { ... }
impl RouteData for Edge { ... }
```

Reply:
Yes, `RouteData` is a better definition. However, if we introduce `RouteData` in `Router`,  then `Router` cannot be made into an object. 
e.g., this would caused a compiling error:
```
#[derive(Clone)]
struct FnGenerator {
    router: Arc<dyn Router>,
}
```
Error:
```
error[E0038]: the trait `Router` cannot be made into an object
  --> runtime/src/assembly.rs:69:17
   |
69 |     router: Arc<dyn Router>,
   |                 ^^^^^^^^^^ `Router` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
`` `